Dashboard should request log streams only when the user gestures to display them

eerhardt opened the pull request:
Change the AppHost to only subscribe to DCP's logs when there is a subscriber for the logs.

Fix #2789
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3235)

davidfowl committed:
WIP

Fix merge issue

davidfowl committed:
Show logs

eerhardt committed:
Merge branch 'main' into on-demand-logs

eerhardt committed:
Remove the "StreamingLogger" because that causes subsequent subscribers to not see what was already written.

Fix duplicate logs issue by clearing out the backlog when the last subscriber leaves.

JamesNK commented:
What happens to historical logs? For example, will requesting the console logs for an app show all longs since the app started? (if less than the limit)

eerhardt committed:
Fix a concurrency issue with the backlog. Ensure the backlog is only snap shotted after subscribing to the log.

Fix existing tests for new functionality and add an additional test.

eerhardt commented:
> What happens to historical logs? For example, will requesting the console logs for an app show all longs since the app started? (if less than the limit)

The "backlog" gets cleared after the last subscriber to a resource leaves. Later, when a "new" subscriber comes along, they are all replayed from DCP because we call `StartLogStream(executable);` again.

So even if no one was subscribed for a while, the whole logs come back once someone does.

davidfowl commented:
One small race is that starting logs before the container is ready, it won’t show up. Super tight race and I don’t know if it’ll be worth complicating the code but I thought I’d let you know

davidfowl commented on src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs:
```diff
@@ -348,6 +331,21 @@ void Log(LogLine log)
+                // to ensure the backlog snapshot contains the correct logs. The backlog
+                // gets cleared when there are no subscribers, so we ensure we are subscribing first.
+                LogLine[] backlogSnapshot;
+                lock (state._backlog)
```
You can’t lock in this method. There’s a compiler bug allowing this currently https://github.com/dotnet/roslyn/issues/72443

eerhardt replied:
I think that bug is different. In that bug it is `yield return`ing from inside the `lock`. Here we aren't doing anything async in the lock, we are just making the snapshot.

I can move this out into a synchronous method (which should obviously be fine being called from an async iterator.

eerhardt committed:
PR feedback

Only clear the backlog on containers and executables.

eerhardt commented:
> One small race is that starting logs before the container is ready, it won’t show up.

Can you describe this a bit more? When what is starting the logs before the container is ready?

eerhardt committed:
PR feedback.

Move lock out of async iterator.

eerhardt committed:
Clean up code.

- Remove count and instead check if the delegate is null to indicate whether there are subscribers or not.
- Remove the separate IAsyncEnumerable classes and just use `async IAsyncEnumerable` methods.

eerhardt committed:
Fix a race at startup when logs aren't available.

Employ a 2nd loop that listens for both "has subscribers" and "logs available".

eerhardt commented:
> One small race is that starting logs before the container is ready, it won’t show up. Super tight race and I don’t know if it’ll be worth complicating the code but I thought I’d let you know

This didn't turn out to be too bad. I added another loop that joined the "has subscribers" and "logs available" information, and only starts the logs once both are true.

eerhardt committed:
Simplify ResourceNotificationService.WatchAsync.

eerhardt committed:
Fix test build

davidfowl commented on tests/Aspire.Hosting.Tests/ResourceNotificationTests.cs:
```diff
@@ -48,7 +48,7 @@ async Task<List<ResourceEvent>> GetValuesAsync(CancellationToken cancellationTok
             var values = new List<ResourceEvent>();
 
-            await foreach (var item in notificationService.WatchAsync().WithCancellation(cancellationToken))
+            await foreach (var item in notificationService.WatchAsync(cancellationToken))
```
Is this right?

davidfowl replied:
Looked at the IL

davidfowl commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -195,6 +199,70 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
+                    (logsAvailable, hasSubscribers) = stateEntry;
+                }
+
+                logsAvailable = entry.LogsAvailable ?? logsAvailable;
```
Once this is true, it'll always to be true, maybe add a comment.

eerhardt committed:
Address PR feedback

- Set SingleReader=true on the logInformationChannel.
- Add comment and assert that LogsAvailable can only turn true and can't go back to false.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into on-demand-logs
